test/src/com/xilinx/rapidwright/util/TestBlackBoxPopulator.java

```diff
@@ -0,0 +1,113 @@
+        add(new Pair<>("hw_contract_pr2", "INT_X0Y0"));
+    }};
+
+    private int numPIPs(Design d) {
+        int PIPCount = 0;
+        for (Net n : d.getNets()) {
+            PIPCount += n.getPIPs().size();
+        }
+        return PIPCount;
+    }
+
+    @Test
+    void testRelocateModuleInsts() {
+        Design top = Design.readCheckpoint(topDCPName);
+        int numCellTop = top.getCells().size();
+        int numVccNetTop = (top.getVccNet() == null) ? 0 : 1;
+        int numGndNetTop = (top.getGndNet() == null) ? 0 : 1;
+        int numSignalNetTop = top.getNets().size() - numVccNetTop - numGndNetTop;
+        int numPIPTop = numPIPs(top);
+
+        Design template = Design.readCheckpoint(cellDCPName);
+        int numVccNetTemplate = (template.getVccNet() == null) ? 0 : 1;
+        int numGndNetTemplate = (template.getGndNet() == null) ? 0 : 1;
+        int numSignalNetTemplate = template.getNets().size() - numVccNetTemplate - numGndNetTemplate;
+        int numPIPTemplate = numPIPs(template);
+
+        Module mod = new Module(template, false);
+        BlackboxPopulator.relocateModuleInsts(top, mod, cellAnchor, targets);
+
+        Assertions.assertEquals(targets.size()*template.getCells().size()+numCellTop, top.getCells().size()
+                ,"Wrong number of cells!");
+
+        int numExpectedNets = targets.size()*numSignalNetTemplate + numSignalNetTop
+                              + Math.max(numVccNetTop, numVccNetTemplate) + Math.max(numGndNetTop, numGndNetTemplate);
+        Assertions.assertEquals(numExpectedNets, top.getNets().size(),"Wrong number of nets!");
+
+        Assertions.assertEquals(targets.size()*numPIPTemplate + numPIPTop, numPIPs(top),"Wrong number of PIPs!");
+    }
+    @Test
+    /**
+     * To catch a problem where the work library appears before its dependencies.
+     *
+     * The current workaround is to put every library into work library,
+     * ie., using consolidateAllToWorkLibrary() at the end of relocateModuleInsts.
+     * This test is specific to that workaround to be fast.
+     * Later when a better solution is adopted, this test will need to adjust to check the solution.
+     */
+    void testOutputDCP() {
+        Design top = Design.readCheckpoint(topDCPName);
+        Design template = Design.readCheckpoint(cellDCPName);
+        Module mod = new Module(template, false);
+        BlackboxPopulator.relocateModuleInsts(top, mod, cellAnchor, targets);
+
+
+        Collection<EDIFLibrary> libs = top.getNetlist().getLibraries();
+        Assertions.assertEquals(2, libs.size(),"Expect two libraries!");
+
+        Set<String> libNames = new HashSet<>();
+        for (EDIFLibrary l : libs) {
+            libNames.add(l.getName());
+        }
+
+        Assertions.assertTrue(libNames.contains("hdi_primitives"),"Expect to have hdi_primitives library!");
+        Assertions.assertTrue(libNames.contains("work"),"Expect to have work library!");
+    }
```

Comment:
Would it make sense to merge this into the test above?

Reply:
combined.